src/mscorlib/src/System/StringComparer.cs

```diff
@@ -213,83 +213,77 @@ internal sealed class CultureAwareComparer : StringComparer
+            x == null ? -1 :
+            y == null ? 1 :
+            string.CompareOrdinal(x, y);
 
-            if( _ignoreCase) {
-                if( x.Length != y.Length) {
-                    return false;
-                }
-                return (String.Compare(x, y, StringComparison.OrdinalIgnoreCase) == 0);                                            
-            }
-            return x.Equals(y);
-        }               
-                
-        public override int GetHashCode(string obj) {
-            if( obj == null) {
-                throw new ArgumentNullException(nameof(obj));
-            }
-            Contract.EndContractBlock();
+        public override bool Equals(string x, string y) =>
+            object.ReferenceEquals(x, y) ||
+            (x != null && x.Equals(y));
 
-            if( _ignoreCase) {
-                return TextInfo.GetHashCodeOrdinalIgnoreCase(obj);
+        public override int GetHashCode(string obj)
+        {
+            if (obj == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.obj);
             }
-                        
             return obj.GetHashCode();
-        }       
-
-        // Equals method for the comparer itself. 
-        public override bool Equals(Object obj){
-            OrdinalComparer comparer = obj as OrdinalComparer;
-            if( comparer == null) {
-                return false;
-            }
-            return (this._ignoreCase == comparer._ignoreCase);
         }
 
-        public override int GetHashCode() {
-            string name = "OrdinalComparer";
-            int hashCode = name.GetHashCode();
-            return _ignoreCase ? (~hashCode) : hashCode;
-        }
+        // Equals/GetHashCode methods for the comparer itself. 
+        public override bool Equals(Object obj) => obj is OrdinalComparer;
+        public override int GetHashCode() => nameof(OrdinalComparer).GetHashCode();
 
 #if FEATURE_RANDOMIZED_STRING_HASHING           
+        IEqualityComparer IWellKnownStringEqualityComparer.GetEqualityComparerForSerialization() => this;
+#endif
+    }
 
-        IEqualityComparer IWellKnownStringEqualityComparer.GetEqualityComparerForSerialization() {
-            return this;
+    // Provide a more optimal implementation of ordinal ignore-case comparison.
+    [Serializable]
+    internal sealed class OrdinalIgnoreCaseComparer : StringComparer
+#if FEATURE_RANDOMIZED_STRING_HASHING
+        , IWellKnownStringEqualityComparer
+#endif
+    {
+        public override int Compare(string x, string y) =>
+            object.ReferenceEquals(x, y) ? 0 :
+            x == null ? -1 :
+            y == null ? 1 :
+            string.Compare(x, y, StringComparison.OrdinalIgnoreCase);
+
+        public override bool Equals(string x, string y) =>
+            object.ReferenceEquals(x, y) ||
+            (x != null && x.Equals(y, StringComparison.OrdinalIgnoreCase));
```

Comment:
Similarly, could this be `public override bool Equals(string x, string y) => string.Equals(x, y, StringComparison.OrdinalIgnoreCase);`?